### PR TITLE
Use `either` crate

### PR DESCRIPTION
**Pull request by Ralith**
I went in planning to remove the internal `Either` type entirely, but obviously we can't implement `Future` for an external type in an external crate. There's a case to be made for removing the internal `Either` type and instead submitting a (feature-guarded?) `Future` impl to the external `Either` crate, but I'm not sure how I feel about that.

I'm also not sure how I feel about the small change here that remains if we're not removing the internal Either`; let's discuss.

### CODE REVIEW

**Comment by Ralith** [comment]
For one thing, the `Either` `Future` impl will be considerably less important once async/await is readily usable, since you'll be able to write branches naturally.

**Comment by cramertj** [comment]
We actually can implement `Future` for the `Either` type in a separate crate-- the impl just has to be inside `futures-core`. I think I'd be +1 for this, but it should be under an `either` feature flag, and we should re-export the type in `futures-util` so that users don't have to manually depend on the `either` crate. WDYT?

**Comment by Ralith** [comment]
I suppose I don't see any reason not to put the impl in `-core`. Would the feature flag be disabled by default?

**Comment by cramertj** [comment]
@Ralith Hm... I think i'd probably disable by default inside futures-core, but make futures-util enable it by default. This way we keep futures-core dependencies as minimal as possible.

File: futures-util/Cargo.toml [hunk 1]
```diff
@@ -12,7 +12,7 @@ Common utilities and extension traits for the futures-rs library.
 
 [features]
 std = ["futures-core/std", "futures-io/std", "futures-sink/std", "either/use_std"]
-default = ["std"]
+default = ["std", "futures-core/either"]
```
Comment (Ralith) [inline futures-util/Cargo.toml]: Is this sufficient, or does the `futures-core` feature need to be handled in other ways?

Comment (cramertj) [reply futures-util/Cargo.toml]: Seems fine to me!

File: futures-util/src/future/mod.rs [hunk 2]
```diff
@@ -46,9 +45,10 @@ pub use self::err_into::ErrInto;
 pub use self::or_else::OrElse;
 pub use self::select::Select;
 pub use self::then::Then;
-pub use self::either::Either;
 pub use self::inspect::Inspect;
 
+pub use either::Either;
```
Comment (cramertj) [inline futures-util/src/future/mod.rs]: Can you also re-export from the `futures` facade crate? Once that's done, I think we can ship this.

File: futures-util/src/lib.rs [hunk 3]
```diff
@@ -9,6 +9,7 @@
 extern crate futures_core;
 extern crate futures_io;
 extern crate futures_sink;
+extern crate either;
```
Comment (kpp) [inline futures-util/src/lib.rs]: `#[cfg(feature = "std")]`?

Comment (Ralith) [reply futures-util/src/lib.rs]: Shouldn't be needed; I'm already carefully dispatching to `either`'s own `std` feature.

File: futures-util/src/future/mod.rs (hunk 2 above)
Comment (Ralith) [reply futures-util/src/future/mod.rs]: Isn't there [a pre-existing export](https://github.com/rust-lang-nursery/futures-rs/blob/master/futures/src/lib.rs#L189)?

File: futures-util/src/future/either.rs [hunk 4]
```diff
@@ -1,55 +0,0 @@
...
-    /// Second branch of the type
-    B(B),
-}
-
-impl<T, A, B> Either<(T, A), (T, B)> {
-    /// Splits out the homogeneous type from an either of tuples.
-    ///
-    /// This method is typically useful when combined with the `Future::select2`
-    /// combinator.
-    pub fn split(self) -> (T, Either<A, B>) {
```
Comment (mathstuf) [inline futures-util/src/future/either.rs]: See #580 for modifications related to this. I don't know if `either` is inclined to add something similar, but if not, the method is useful enough that free-standing functions may still be useful.

File: futures-util/src/future/mod.rs (hunk 2 above)
Comment (cramertj) [reply futures-util/src/future/mod.rs]: Oh, you're right! Sorry, I that didn't occur to me.

**Comment by cramertj** [comment]
Rebased and reopened https://github.com/rust-lang-nursery/futures-rs/pull/828 targeting master.

**Pull request closed without merging**